### ISSUE TITLE
False positive: Name mangling + `@dataclass` (undefined_attribute)

### ISSUE THREAD

**Issue by mflova** [quora/pyanalyze#657]
This is a very specific issue. So maybe it is not needed to be fixed as I doubt somebody will ever go through this situation. Minimum reproducible example:

```python
from dataclasses import dataclass

@dataclass
class A:
    def __func(self) -> None:
        ...

    def run(self) -> None:
        self.__func()  # ERROR: Undefined attribute
```

This issue only happens with `@dataclass`. If the decorator is removed, it will work fine.

**Comment by JelleZijlstra** [comment]
Thanks, this actually indicates a more serious bug: the attribute logic doesn't perform name mangling. You see it only with dataclasses because for other classes pyanalyze is conservative and doesn't necessarily report undefined attributes. I'll fix this.

**Comment by mflova** [comment]
I was wondering the same. If `pyanalyze` is able to detect the underlying logic behind `dataclass` decorator, having this issue only with `@dataclass` was strange to me
